marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java

```diff
@@ -59,7 +62,59 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
         @Override
         public void awaitCompletion() {
             logger.trace("output endpoint running endpoint={} work={}", getEndpointPath(), getWorkUnit());
+            setPhase(WorkPhase.RUNNING);
+            SessionState session = allowsSession() ? getEndpoint().getCaller().newSessionState() : null;
 
+            calling: while (true) {
+                Stream<InputStream> output = null;
+                try {
+                    logger.trace("output endpoint={} count={} state={}",
+                            getEndpointPath(), getCallCount(), getEndpointState());
+
+                    output = getEndpoint().getCaller().call(
+                            getEndpoint().getClient(), getEndpointState(), session, getWorkUnit()
+                    );
+                    incrementCallCount();
+                } catch(Throwable throwable) {
+                    throw new RuntimeException("error while calling "+getEndpoint().getEndpointPath(), throwable);
+                }
+
+                if (allowsEndpointState()) {
+                    if(output!= null && output.findFirst()!= null && output.findFirst().isPresent()) {
```

Comment:
findFirst() is a terminal operation that completely consumes the stream.

Would it work to perform the tests in steps?

That is, if the output is not null, construct the array.  Then, if the array is not empty, process it?